api/ApiRouter.inc

```diff
@@ -59,18 +59,22 @@ class ApiRouter
         $node = $this->root;
         $data = [];
         $parts = explode("/", $url);
-        foreach ($parts as $part) {
-            $next_node = $node->children[$part] ?? null;
-            if ($next_node) {
-                $node = $next_node;
-            } else {
-                [$param_name, $validator] = $this->get_validator($node);
-                $node = $node->children[$param_name];
-                $data[$param_name] = $validator($part, $data);
+        try {
+            foreach ($parts as $part) {
+                $next_node = $node->children[$part] ?? null;
+                if ($next_node) {
+                    $node = $next_node;
+                } else {
+                    [$param_name, $validator] = $this->get_validator($node);
+                    $node = $node->children[$param_name];
+                    $data[$param_name] = $validator($part, $data);
+                }
             }
-        }
-        if (empty($node->handlers)) {
-            throw new InvalidAPI();
+            if (empty($node->handlers)) {
+                throw new InvalidAPI();
+            }
+        } catch (InvalidAPI $exception) {
+            throw new InvalidAPI($exception->getMessage() . ": $url");
```

Comment:
While I agree with Casey that this is PR is unneeded (and seems like something off with Firefox, maybe you need to catch the socket when it is created and before Firefox starts reusing it (a bit like how websockets need to be handled)), I wonder if we couldn't help with the issue in a way that's more useful in general.

Something that may have helped clear some confusion is logging the URL in our error logs when we're throwing an exception in the API layer. We could handle this by improving our existing global exception handler to call `error_log` with some API specific information (like actual api called, ...). Thoughts?

Reply:
`error_log` is going to be putting the data in the wrong place to be useful. The API user is the one who might benefit from this knowledge, but it'll be logged server-side which virtually none of them will be able to access -- Ray on the TEST being the exception because he's a developer.

Doing this temporarily in a sandbox to help debug things during development is a great idea though.

Reply:
(I didn't know I could do that)

Reply:
Oh sure, you're welcome to use error_log during development. There's a sandbox-specific php_errors file that sandboxes dump into.